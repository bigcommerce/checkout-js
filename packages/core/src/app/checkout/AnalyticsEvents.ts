let boltTracker: Window["BoltTrack"] = {
  recordEvent: (event: string) => console.log("--bolt bigc--: NOOP ", event),
};

const eventLog: string[] = [];

export const AnalyticsEvents = {
  init: () => {
    try {
      console.log("--bolt bigc--: intializing Bolt-BigC analytics events");
      if (window && window.BoltTrack) {
        boltTracker = window.BoltTrack;
        console.log("--bolt bigc--: successfully assigned window BoltTrack");
        // immediately send checkout load success event
        AnalyticsEvents.emitEvent("Checkout load success");
      }
    } catch (e) {
      console.log("--bolt bigc--: error during emitting event ", e);
    }
  },
  emitEvent: (eventName: string) => {
    try {
      console.log("--bolt bigc--: emitting analytics event ", eventName);
      const props: any = {
        nextState: eventName,
        prevState: "",
      };
      if (eventLog.length > 0) {
        props.prevState = eventLog[0];
      }
      boltTracker.recordEvent("CheckoutFunnelTransition", props);
      // add latest event to beginning of log array
      eventLog.unshift(eventName);
    } catch (e) {
      console.log("--bolt bigc--: error during emitting event ", e);
    }
  },
  onBeforeUnload: () => {
    AnalyticsEvents.emitEvent("Exit");
  },
};
