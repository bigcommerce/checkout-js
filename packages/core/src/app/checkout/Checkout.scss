@import '../ui/Base';

$color-secondary: #757575;
$color-link: #476BEF;
$fontFamily-montserrat: 'Montserrat';

// optimizedCheckout-headingPrimary
.header, .header-secondary, .sub-header {
    color: $color-black;
    font-family: $fontFamily-montserrat;
    font-size: fontSize("large");
    font-weight: fontWeight("bold");
}

.header-secondary {
    color: $color-secondary;
}

// optimizedCheckout-headingSecondary
.sub-header {
    font-size: fontSize("smaller");
}

// optimizedCheckout-contentPrimary
.body-regular, .body-cta, .body-medium, .body-bold {
    color: $color-black;
    font-family: $fontFamily-montserrat;
    font-size: fontSize("base");
    font-weight: fontWeight("normal");
}

.body-cta {
    color: $color-link;
}

.body-medium, .body-bold {
    font-weight: fontWeight("medium");
}

.body-bold {
    font-size: fontSize("base");
    font-weight: fontWeight("bold");
}

.sub-text, .sub-text-medium, .sub-text-bold {
    color: $color-black;
    font-family: $fontFamily-montserrat;
    font-size: fontSize("smallest");
    font-weight: fontWeight("normal");
}

.sub-text-medium {
    color: $color-secondary;
    font-weight: fontWeight("medium");
}

.sub-text-bold {
    font-weight: fontWeight("bold");
}

// overrides 
.body-bold, .body-medium {
    &.button--primary:not(:disabled){
        color: $color-white;
    }
}

.form-checkbox + .form-label,
.form-radio + .form-label {
    &.body-regular {
        color: $color-black;
        font-family: $fontFamily-montserrat;
        font-size: fontSize("base");
        font-weight: fontWeight("normal");
    }
}

.shippingOptionLabel {
    .shippingOption-price, .shippingOption-desc {
        &.body-medium {
            color: $color-black;
            font-family: $fontFamily-montserrat;
            font-size: fontSize("base");
            font-weight: fontWeight("medium");
        }
    }
}

.body-medium {
     small {
        color: $color-black;
        font-family: $fontFamily-montserrat;
        font-size: fontSize("base");
        font-weight: fontWeight("medium");
    }
}

.add-consignment-button {
    &.body-regular {
        color: $color-black;
    }
}
