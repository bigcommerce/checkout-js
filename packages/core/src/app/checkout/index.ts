export { RenderCheckout, RenderCheckoutOptions } from './renderCheckout';
export { default as CheckoutContext, CheckoutContextProps } from './CheckoutContext';
export {
    default as CheckoutProvider,
    CheckoutProviderProps,
    CheckoutProviderState,
} from './CheckoutProvider';
export { default as withCheckout, WithCheckoutProps } from './withCheckout';
export { default as CheckoutSupport } from './CheckoutSupport';
export { default as NoopCheckoutSupport } from './NoopCheckoutSupport';
export { default as navigateToOrderConfirmation } from './navigateToOrderConfirmation';
