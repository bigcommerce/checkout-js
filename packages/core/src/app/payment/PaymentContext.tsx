import { PaymentMethod } from '@bigcommerce/checkout-sdk';
import { createContext } from 'react';
import { ObjectSchema } from 'yup';

import { PaymentFormValues } from '@bigcommerce/checkout/payment-integration-api';

export interface PaymentContextProps {
    disableSubmit(method: PaymentMethod, disabled?: boolean): void;
    // NOTE: This prop allows certain payment methods to override the default
    // form submission behaviour. It is not recommended to use it because
    // generally speaking we want to avoid method-specific snowflake behaviours.
    // Nevertheless, because of some product / UX decisions made in the past
    // (i.e.: Amazon), we have to have this backdoor so we can preserve these
    // snowflake behaviours. In the future, if we decide to change the UX, we
    // can remove this prop.
    setSubmit(method: PaymentMethod, fn: ((values: PaymentFormValues) => void) | null): void;
    setValidationSchema(
        method: PaymentMethod,
        schema: ObjectSchema<Partial<PaymentFormValues>> | null,
    ): void;
    hidePaymentSubmitButton(method: PaymentMethod, hidden?: boolean): void;
}

const PaymentContext = createContext<PaymentContextProps | undefined>(undefined);

export default PaymentContext;
