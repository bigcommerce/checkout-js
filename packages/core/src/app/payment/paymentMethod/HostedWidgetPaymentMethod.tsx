import {
    AccountInstrument,
    CardInstrument,
    CheckoutSelectors,
    CustomerInitializeOptions,
    CustomerRequestOptions,
    Instrument,
    PaymentInitializeOptions,
    PaymentInstrument,
    PaymentMethod,
    PaymentRequestOptions,
} from '@bigcommerce/checkout-sdk';
import { memoizeOne } from '@bigcommerce/memoize';
import classNames from 'classnames';
import { find, noop, some } from 'lodash';
import React, { Component, ReactNode } from 'react';
import { ObjectSchema } from 'yup';

import { PaymentFormValues } from '@bigcommerce/checkout/payment-integration-api';

import { CheckoutContextProps, withCheckout } from '../../checkout';
import { preventDefault } from '../../common/dom';
import { connectFormik, ConnectFormikProps } from '../../common/form';
import { MapToPropsFactory } from '../../common/hoc';
import { TranslatedString } from '../../locale';
import { LoadingOverlay } from '../../ui/loading';
import {
    AccountInstrumentFieldset,
    CardInstrumentFieldset,
    isBankAccountInstrument,
    isCardInstrument,
    isInstrumentCardCodeRequiredSelector,
    isInstrumentCardNumberRequiredSelector,
    isInstrumentFeatureAvailable,
} from '../storedInstrument';
import StoreInstrumentFieldset from '../StoreInstrumentFieldset';
import withPayment, { WithPaymentProps } from '../withPayment';

import SignOutLink from './SignOutLink';

export interface HostedWidgetPaymentMethodProps {
    additionalContainerClassName?: string;
    buttonId?: string;
    containerId: string;
    hideContentWhenSignedOut?: boolean;
    hideVerificationFields?: boolean;
    isAccountInstrument?: boolean;
    hideWidget?: boolean;
    isInitializing?: boolean;
    isUsingMultiShipping?: boolean;
    isSignInRequired?: boolean;
    method: PaymentMethod;
    paymentDescriptor?: string;
    shouldHideInstrumentExpiryDate?: boolean;
    shouldShow?: boolean;
    shouldShowDescriptor?: boolean;
    shouldShowEditButton?: boolean;
    shouldRenderCustomInstrument?: boolean;
    storedCardValidationSchema?: ObjectSchema;
    renderCustomPaymentForm?(): React.ReactNode;
    validateInstrument?(
        shouldShowNumberField: boolean,
        selectedInstrument?: CardInstrument,
    ): React.ReactNode;
    deinitializeCustomer?(options: CustomerRequestOptions): Promise<CheckoutSelectors>;
    deinitializePayment(options: PaymentRequestOptions): Promise<CheckoutSelectors>;
    initializeCustomer?(options: CustomerInitializeOptions): Promise<CheckoutSelectors>;
    initializePayment(
        options: PaymentInitializeOptions,
        selectedInstrument?: CardInstrument,
    ): Promise<CheckoutSelectors>;
    onPaymentSelect?(): void;
    onSignOut?(): void;
    onSignOutError?(error: Error): void;
    onUnhandledError?(error: Error): void;
    signInCustomer?(): void;
}

interface WithCheckoutHostedWidgetPaymentMethodProps {
    instruments: PaymentInstrument[];
    isInstrumentFeatureAvailable: boolean;
    isLoadingInstruments: boolean;
    isPaymentDataRequired: boolean;
    isSignedIn: boolean;
    isInstrumentCardCodeRequired(instrument: Instrument, method: PaymentMethod): boolean;
    isInstrumentCardNumberRequired(instrument: Instrument): boolean;
    loadInstruments(): Promise<CheckoutSelectors>;
    signOut(options: CustomerRequestOptions): void;
}

interface HostedWidgetPaymentMethodState {
    isAddingNewCard: boolean;
    selectedInstrumentId?: string;
}

class HostedWidgetPaymentMethod extends Component<
    HostedWidgetPaymentMethodProps &
        WithCheckoutHostedWidgetPaymentMethodProps &
        ConnectFormikProps<PaymentFormValues> &
        WithPaymentProps
> {
    state: HostedWidgetPaymentMethodState = {
        isAddingNewCard: false,
    };

    async componentDidMount(): Promise<void> {
        const {
            isInstrumentFeatureAvailable: isInstrumentFeatureAvailableProp,
            loadInstruments,
            method,
            onUnhandledError = noop,
            setValidationSchema,
        } = this.props;

        setValidationSchema(method, this.getValidationSchema());

        try {
            if (isInstrumentFeatureAvailableProp) {
                await loadInstruments();
            }

            await this.initializeMethod();
        } catch (error) {
            onUnhandledError(error);
        }
    }

    async componentDidUpdate(
        prevProps: Readonly<
            HostedWidgetPaymentMethodProps & WithCheckoutHostedWidgetPaymentMethodProps
        >,
        prevState: Readonly<HostedWidgetPaymentMethodState>,
    ): Promise<void> {
        const {
            deinitializePayment = noop,
            instruments,
            method,
            onUnhandledError = noop,
            setValidationSchema,
        } = this.props;

        const { selectedInstrumentId } = this.state;

        setValidationSchema(method, this.getValidationSchema());

        if (
            selectedInstrumentId !== prevState.selectedInstrumentId ||
            (prevProps.instruments.length > 0 && instruments.length === 0)
        ) {
            try {
                await deinitializePayment({
                    gatewayId: method.gateway,
                    methodId: method.id,
                });
                await this.initializeMethod();
            } catch (error) {
                onUnhandledError(error);
            }
        }
    }

    async componentWillUnmount(): Promise<void> {
        const {
            deinitializeCustomer = noop,
            deinitializePayment = noop,
            method,
            onUnhandledError = noop,
            setSubmit,
            setValidationSchema,
        } = this.props;

        setValidationSchema(method, null);
        setSubmit(method, null);

        try {
            await deinitializePayment({
                gatewayId: method.gateway,
                methodId: method.id,
            });

            await deinitializeCustomer({
                methodId: method.id,
            });
        } catch (error) {
            onUnhandledError(error);
        }
    }

    render(): ReactNode {
        const {
            instruments,
            hideWidget = false,
            isInitializing = false,
            isSignedIn = false,
            method,
            isAccountInstrument,
            isInstrumentFeatureAvailable: isInstrumentFeatureAvailableProp,
            isLoadingInstruments,
            shouldHideInstrumentExpiryDate = false,
            shouldShow = true,
        } = this.props;

        const { isAddingNewCard, selectedInstrumentId = this.getDefaultInstrumentId() } =
            this.state;

        if (!shouldShow) {
            return null;
        }

        const selectedInstrument =
            instruments.find((instrument) => instrument.bigpayToken === selectedInstrumentId) ||
            instruments[0];

        const shouldShowInstrumentFieldset =
            isInstrumentFeatureAvailableProp && instruments.length > 0;
        const shouldShowCreditCardFieldset = !shouldShowInstrumentFieldset || isAddingNewCard;
        const isLoading = (isInitializing || isLoadingInstruments) && !hideWidget;

        const selectedAccountInstrument = this.getSelectedBankAccountInstrument(
            isAddingNewCard,
            selectedInstrument,
        );
        const shouldShowAccountInstrument =
            instruments[0] && isBankAccountInstrument(instruments[0]);

        return (
            <LoadingOverlay hideContentWhenLoading isLoading={isLoading}>
                <div className="paymentMethod--hosted">
                    {shouldShowAccountInstrument && shouldShowInstrumentFieldset && (
                        <AccountInstrumentFieldset
                            instruments={instruments as AccountInstrument[]}
                            onSelectInstrument={this.handleSelectInstrument}
                            onUseNewInstrument={this.handleUseNewCard}
                            selectedInstrument={selectedAccountInstrument}
                        />
                    )}

                    {!shouldShowAccountInstrument && shouldShowInstrumentFieldset && (
                        <CardInstrumentFieldset
                            instruments={instruments as CardInstrument[]}
                            onDeleteInstrument={this.handleDeleteInstrument}
                            onSelectInstrument={this.handleSelectInstrument}
                            onUseNewInstrument={this.handleUseNewCard}
                            selectedInstrumentId={selectedInstrumentId}
                            shouldHideExpiryDate={shouldHideInstrumentExpiryDate}
                            validateInstrument={this.getValidateInstrument()}
                        />
                    )}

                    {this.renderPaymentDescriptorIfAvailable()}

                    {this.renderContainer(shouldShowCreditCardFieldset)}

                    {isInstrumentFeatureAvailableProp && (
                        <StoreInstrumentFieldset
                            instrumentId={selectedInstrumentId}
                            isAccountInstrument={isAccountInstrument || shouldShowAccountInstrument}
                        />
                    )}

                    {this.renderEditButtonIfAvailable()}

                    {isSignedIn && <SignOutLink method={method} onSignOut={this.handleSignOut} />}
                </div>
            </LoadingOverlay>
        );
    }

    getValidateInstrument(): ReactNode {
        const {
            hideVerificationFields,
            instruments,
            isInstrumentCardNumberRequired: isInstrumentCardNumberRequiredProp,
            validateInstrument,
        } = this.props;

        const { selectedInstrumentId = this.getDefaultInstrumentId() } = this.state;
        const selectedInstrument = find(instruments, {
            bigpayToken: selectedInstrumentId,
        }) as CardInstrument;
        const shouldShowNumberField = selectedInstrument
            ? isInstrumentCardNumberRequiredProp(selectedInstrument)
            : false;

        if (hideVerificationFields) {
            return;
        }

        if (validateInstrument) {
            return validateInstrument(shouldShowNumberField, selectedInstrument);
        }
    }

    renderContainer(shouldShowCreditCardFieldset: any): ReactNode {
        const {
            containerId,
            hideContentWhenSignedOut = false,
            hideWidget,
            isSignInRequired = false,
            isSignedIn,
            method,
            additionalContainerClassName,
            shouldRenderCustomInstrument = false,
            renderCustomPaymentForm,
        } = this.props;

        return (
            <div
                className={classNames(
                    'widget',
                    `widget--${method.id}`,
                    'payment-widget',
                    shouldRenderCustomInstrument ? '' : additionalContainerClassName,
                )}
                id={containerId}
                style={{
                    display:
                        (hideContentWhenSignedOut && isSignInRequired && !isSignedIn) ||
                        !shouldShowCreditCardFieldset ||
                        hideWidget
                            ? 'none'
                            : undefined,
                }}
                tabIndex={-1}
            >
                {shouldRenderCustomInstrument &&
                    renderCustomPaymentForm &&
                    renderCustomPaymentForm()}
            </div>
        );
    }

    private getValidationSchema(): ObjectSchema | null {
        const {
            isInstrumentFeatureAvailable: isInstrumentFeatureAvailableProp,
            isPaymentDataRequired,
            storedCardValidationSchema,
        } = this.props;

        if (!isPaymentDataRequired) {
            return null;
        }

        const selectedInstrument = this.getSelectedInstrument();

        if (isInstrumentFeatureAvailableProp && selectedInstrument) {
            return storedCardValidationSchema || null;
        }

        return null;
    }

    private getSelectedInstrument(): PaymentInstrument | undefined {
        const { instruments } = this.props;
        const { selectedInstrumentId = this.getDefaultInstrumentId() } = this.state;

        return find(instruments, { bigpayToken: selectedInstrumentId });
    }

    private handleDeleteInstrument: (id: string) => void = (id) => {
        const {
            instruments,
            formik: { setFieldValue },
        } = this.props;
        const { selectedInstrumentId } = this.state;

        if (instruments.length === 0) {
            this.setState({
                isAddingNewCard: true,
                selectedInstrumentId: undefined,
            });

            setFieldValue('instrumentId', '');
        } else if (selectedInstrumentId === id) {
            this.setState({
                selectedInstrumentId: this.getDefaultInstrumentId(),
            });

            setFieldValue('instrumentId', this.getDefaultInstrumentId());
        }
    };

    private getSelectedBankAccountInstrument(
        isAddingNewCard: boolean,
        selectedInstrument: PaymentInstrument,
    ): AccountInstrument | undefined {
        return !isAddingNewCard && selectedInstrument && isBankAccountInstrument(selectedInstrument)
            ? selectedInstrument
            : undefined;
    }

    private renderEditButtonIfAvailable() {
        const { shouldShowEditButton, buttonId } = this.props;
        const translatedString = <TranslatedString id="remote.select_different_card_action" />;

        if (shouldShowEditButton) {
            return (
                <p>
                    <a
                        className={classNames('stepHeader', 'widget-link-amazonpay')}
                        id={buttonId}
                        onClick={preventDefault()}
                    >
                        {translatedString}
                    </a>
                </p>
            );
        }
    }

    private renderPaymentDescriptorIfAvailable() {
        const { shouldShowDescriptor, paymentDescriptor } = this.props;

        if (shouldShowDescriptor && paymentDescriptor) {
            return <div className="payment-descriptor">{paymentDescriptor}</div>;
        }
    }

    private async initializeMethod(): Promise<CheckoutSelectors | void> {
        const {
            isPaymentDataRequired,
            isSignedIn,
            isSignInRequired,
            initializeCustomer = noop,
            initializePayment = noop,
            instruments,
            method,
            setSubmit,
            signInCustomer = noop,
        } = this.props;

        const { selectedInstrumentId = this.getDefaultInstrumentId() } = this.state;

        if (!isPaymentDataRequired) {
            setSubmit(method, null);

            return Promise.resolve();
        }

        if (isSignInRequired && !isSignedIn) {
            setSubmit(method, signInCustomer);

            return initializeCustomer({
                methodId: method.id,
            });
        }

        setSubmit(method, null);

        const selectedInstrument =
            instruments.find((instrument) => instrument.bigpayToken === selectedInstrumentId) ||
            instruments[0];

        return initializePayment(
            {
                gatewayId: method.gateway,
                methodId: method.id,
            },
            selectedInstrument,
        );
    }

    private getDefaultInstrumentId(): string | undefined {
        const { isAddingNewCard } = this.state;

        if (isAddingNewCard) {
            return;
        }

        const { instruments } = this.props;
        const defaultInstrument =
            instruments.find((instrument) => instrument.defaultInstrument) || instruments[0];

        return defaultInstrument && defaultInstrument.bigpayToken;
    }

    private handleUseNewCard: () => void = async () => {
        const { deinitializePayment = noop, initializePayment = noop, method } = this.props;

        this.setState({
            isAddingNewCard: true,
            selectedInstrumentId: undefined,
        });

        await deinitializePayment({
            gatewayId: method.gateway,
            methodId: method.id,
        });

        await initializePayment({
            gatewayId: method.gateway,
            methodId: method.id,
        });
    };

    private handleSelectInstrument: (id: string) => void = (id) => {
        this.setState({
            isAddingNewCard: false,
            selectedInstrumentId: id,
        });
    };

    private handleSignOut: () => void = async () => {
        const { method, onSignOut = noop, onSignOutError = noop, signOut } = this.props;

        try {
            await signOut({ methodId: method.id });
            onSignOut();
        } catch (error) {
            onSignOutError(error);
        }
    };
}

const mapFromCheckoutProps: MapToPropsFactory<
    CheckoutContextProps,
    WithCheckoutHostedWidgetPaymentMethodProps,
    HostedWidgetPaymentMethodProps & ConnectFormikProps<PaymentFormValues>
> = () => {
    const filterInstruments = memoizeOne((instruments: PaymentInstrument[] = []) =>
        instruments.filter(
            (instrument) => isCardInstrument(instrument) || isBankAccountInstrument(instrument),
        ),
    );

    return (context, props) => {
        const { isUsingMultiShipping = false, method } = props;

        const { checkoutService, checkoutState } = context;

        const {
            data: { getCheckout, getConfig, getCustomer, getInstruments, isPaymentDataRequired },
            statuses: { isLoadingInstruments },
        } = checkoutState;

        const checkout = getCheckout();
        const config = getConfig();
        const customer = getCustomer();

        if (!checkout || !config || !customer || !method) {
            return null;
        }

        return {
            instruments: filterInstruments(getInstruments(method)),
            isLoadingInstruments: isLoadingInstruments(),
            isPaymentDataRequired: isPaymentDataRequired(),
            isSignedIn: some(checkout.payments, { providerId: method.id }),
            isInstrumentCardCodeRequired: isInstrumentCardCodeRequiredSelector(checkoutState),
            isInstrumentCardNumberRequired: isInstrumentCardNumberRequiredSelector(checkoutState),
            isInstrumentFeatureAvailable: isInstrumentFeatureAvailable({
                config,
                customer,
                isUsingMultiShipping,
                paymentMethod: method,
            }),
            loadInstruments: checkoutService.loadInstruments,
            signOut: checkoutService.signOutCustomer,
        };
    };
};

export default connectFormik(
    withPayment(withCheckout(mapFromCheckoutProps)(HostedWidgetPaymentMethod)),
);
