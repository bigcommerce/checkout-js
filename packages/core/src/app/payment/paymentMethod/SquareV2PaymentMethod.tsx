import { PaymentInitializeOptions } from '@bigcommerce/checkout-sdk';
import { difference } from 'lodash';
import React, { FunctionComponent, useCallback, useContext, useEffect, useState } from 'react';

import { getAppliedStyles } from '../../common/dom';
import PaymentContext from '../PaymentContext';

import HostedWidgetPaymentMethod , { HostedWidgetPaymentMethodProps } from './HostedWidgetPaymentMethod';

export type SquareV2PaymentMethodProps = Omit<HostedWidgetPaymentMethodProps, 'containerId'>;

const SquareV2PaymentMethod: FunctionComponent<SquareV2PaymentMethodProps> = ({
    initializePayment,
    method,
    ...rest
}) => {
    const [disabled, setDisabled] = useState(true);
    const paymentContext = useContext(PaymentContext);

    useEffect(() => {
        paymentContext?.disableSubmit(method, disabled);

        return () => paymentContext?.disableSubmit(method, false);
    }, [paymentContext, method, disabled]);

    const onValidationChange = useCallback((isValid: boolean) => setDisabled(!isValid), [setDisabled]);
    const renderDummyFormFields = () => {
        return (
            <div data-test="squarev2_dummy_form" style={ { display: 'none' } }>
                <div className="form-field">
                    <div className="form-label optimizedCheckout-form-label" id="messageIsDefault" />
                    <div className="form-input optimizedCheckout-form-input" id="inputIsDefault" />
                </div>
                <div className="form-field">
                    <div className="form-input optimizedCheckout-form-input form-input--focus optimizedCheckout-form-input--focus" id="inputIsFocus" />
                </div>
                <div className="form-field form-field--error">
                    <div className="form-inlineMessage" id="messageIsError" />
                    <div className="form-input optimizedCheckout-form-input" id="inputIsError" />
                </div>
            </div>
        );
    };
    const getStylesFromElement = (id: string, properties: string[]) => {
        const container = document.querySelector<HTMLDivElement>(`#${id}`);

        if (!container) {
            throw new Error(
                `Unable to retrieve input styles as the provided container ID "${id}" is not valid.`,
            );
        }

        return getAppliedStyles(container, properties);
    };
    const copyStylesFromDummyFormFields = () => {
        const styleProps = ['backgroundColor', 'borderColor', 'borderRadius', 'borderWidth', 'color', 'fontSize', 'fontWeight'];

        return {
            default: {
                message: getStylesFromElement('messageIsDefault', [ 'color' ]),
                input: getStylesFromElement('inputIsDefault', styleProps),
            },
            focus: {
                input: getStylesFromElement('inputIsFocus', difference(styleProps, ['borderRadius'])),
            },
            error: {
                message: getStylesFromElement('messageIsError', [ 'color' ]),
                input: getStylesFromElement('inputIsError', [ 'borderColor', 'borderWidth', 'color' ]),
            },
        };
    };
    const mapToSquareStyles = (styles: ReturnType<typeof copyStylesFromDummyFormFields>) => {
        const { borderColor, borderRadius, borderWidth, ...input } = styles.default.input;
        const { borderColor: borderColorIsFocus, borderWidth: borderWidthIsFocus, ...inputIsFocus } = styles.focus.input;
        const { borderColor: borderColorIsError, borderWidth: borderWidthIsError, color: colorIsError } = styles.error.input;

        return {
            input,
            'input.is-focus': {
                ...inputIsFocus,
            },
            'input.is-error': {
                color: colorIsError,
            },
            '.input-container': {
                borderColor,
                borderRadius,
                borderWidth,
            },
            '.input-container.is-focus': {
                borderColor: borderColorIsFocus,
                borderWidth: borderWidthIsFocus,
            },
            '.input-container.is-error': {
                borderColor: borderColorIsError,
                borderWidth: borderWidthIsError,
            },
            '.message-text': {
                color: styles.default.message.color,
            },
            '.message-icon': {
                color: styles.default.message.color,
            },
            '.message-text.is-error': {
                color: styles.error.message.color,
            },
            '.message-icon.is-error': {
                color: styles.error.message.color,
            }
        };
    };

    const containerId = 'squarev2_payment_element_container';
    const initializeSquareV2Payment = useCallback(
        (options: PaymentInitializeOptions) =>
            initializePayment({
                ...options,
                squarev2: {
                    containerId,
                    style: mapToSquareStyles(copyStylesFromDummyFormFields()),
                    onValidationChange,
                },
            }),
        [initializePayment]
    );

    return (
        <>
            { renderDummyFormFields() }
            <HostedWidgetPaymentMethod
                { ...rest }
                containerId={ containerId }
                initializePayment={ initializeSquareV2Payment }
                method={ method }
            />
        </>
    );
};

export default SquareV2PaymentMethod;
