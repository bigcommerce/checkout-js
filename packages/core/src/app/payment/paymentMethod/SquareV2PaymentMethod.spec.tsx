import {
    CheckoutService,
    createCheckoutService,
    PaymentInitializeOptions,
    PaymentMethod,
} from '@bigcommerce/checkout-sdk';
import { mount } from 'enzyme';
import { Formik } from 'formik';
import { noop } from 'lodash';
import React, { FunctionComponent } from 'react';
import { act } from 'react-dom/test-utils';

import { CheckoutProvider } from '../../checkout';
import { getPaymentMethod } from '../payment-methods.mock';
import PaymentContext, { PaymentContextProps } from '../PaymentContext';

import HostedWidgetPaymentMethod from './HostedWidgetPaymentMethod';
import SquareV2PaymentMethod, { SquareV2PaymentMethodProps } from './SquareV2PaymentMethod';

describe('when using SquareV2PaymentMethod', () => {
    let checkoutService: CheckoutService;
    let paymentContext: PaymentContextProps;
    let method: PaymentMethod;
    let squareProps: SquareV2PaymentMethodProps;
    let SquareV2PaymentMethodTest: FunctionComponent;

    beforeEach(() => {
        checkoutService = createCheckoutService();
        paymentContext = {
            disableSubmit: jest.fn(),
            setSubmit: jest.fn(),
            setValidationSchema: jest.fn(),
            hidePaymentSubmitButton: jest.fn(),
        };

        method = {
            ...getPaymentMethod(),
            id: 'squarev2',
        };
        squareProps = {
            method,
            deinitializePayment: jest.fn(),
            initializePayment: jest.fn(),
        };

        const dummyElement = document.createElement('div');

        dummyElement.style.backgroundColor = 'rgb(1, 1, 1)';
        dummyElement.style.borderColor = 'rgb(2, 2, 2)';
        dummyElement.style.borderRadius = '3px';
        dummyElement.style.borderWidth = '4px';
        dummyElement.style.color = 'rgb(5, 5, 5)';
        dummyElement.style.fontSize = '6px';
        dummyElement.style.fontWeight = 'normal';

        jest.spyOn(document, 'querySelector').mockReturnValue(dummyElement);

        SquareV2PaymentMethodTest = () => (
            <CheckoutProvider checkoutService={checkoutService}>
                <PaymentContext.Provider value={paymentContext}>
                    <Formik initialValues={{}} onSubmit={noop}>
                        <SquareV2PaymentMethod {...squareProps} />
                    </Formik>
                </PaymentContext.Provider>
            </CheckoutProvider>
        );
    });

    it('should render dummy form fields', () => {
        const dummyForm = mount(<SquareV2PaymentMethodTest />).find(
            '[data-test="squarev2_dummy_form"]',
        );

        expect(dummyForm).toMatchSnapshot();
    });

    it('renders as HostedWidgetPaymentMethod', () => {
        const hostedWidgetProps = mount(<SquareV2PaymentMethodTest />)
            .find(HostedWidgetPaymentMethod)
            .props();

        expect(hostedWidgetProps).toEqual(
            expect.objectContaining({
                containerId: 'squarev2_payment_element_container',
                initializePayment: expect.any(Function),
                method,
            })
        );
    });

    it('initializes method with required config', async () => {
        const { id: methodId, gateway: gatewayId } = method;
        const initializePayment = mount(<SquareV2PaymentMethodTest />)
            .find(HostedWidgetPaymentMethod)
            .prop('initializePayment');

        await initializePayment({
            methodId,
            gatewayId,
        });

        expect(squareProps.initializePayment).toHaveBeenCalledWith(
            expect.objectContaining({
                methodId: 'squarev2',
                squarev2: {
                    containerId: 'squarev2_payment_element_container',
                    style: {
                        input: {
                            backgroundColor: 'rgb(1, 1, 1)',
                            color: 'rgb(5, 5, 5)',
                            fontSize: '6px',
                            fontWeight: 'normal'
                        },
                        'input.is-focus': {
                            backgroundColor: 'rgb(1, 1, 1)',
                            color: 'rgb(5, 5, 5)',
                            fontSize: '6px',
                            fontWeight: 'normal'
                        },
                        'input.is-error': { color: 'rgb(5, 5, 5)' },
                        '.input-container': {
                            borderColor: 'rgb(2, 2, 2)',
                            borderRadius: '3px',
                            borderWidth: '4px'
                        },
                        '.input-container.is-focus': { borderColor: 'rgb(2, 2, 2)', borderWidth: '4px' },
                        '.input-container.is-error': { borderColor: 'rgb(2, 2, 2)', borderWidth: '4px' },
                        '.message-text': { color: 'rgb(5, 5, 5)' },
                        '.message-icon': { color: 'rgb(5, 5, 5)' },
                        '.message-text.is-error': { color: 'rgb(5, 5, 5)' },
                        '.message-icon.is-error': { color: 'rgb(5, 5, 5)' }
                    },
                    onValidationChange: expect.any(Function),
                },
            }),
        );
    });
        
    it('should enable submit button', async () => {
        const { id: methodId, gateway: gatewayId } = method;
        const initializePayment = mount(<SquareV2PaymentMethodTest />)
            .find(HostedWidgetPaymentMethod)
            .prop('initializePayment');

        await initializePayment({
            methodId,
            gatewayId,
        });

        const {
            squarev2: { onValidationChange },
        } = (squareProps.initializePayment as jest.Mock).mock
            .calls[0][0] as PaymentInitializeOptions['squarev2'];

        act(() => {
            onValidationChange(true);
        });

        expect(paymentContext.disableSubmit).toHaveBeenNthCalledWith(1, method, true);
        expect(paymentContext.disableSubmit).toHaveBeenNthCalledWith(2, method, false);
    });
});
