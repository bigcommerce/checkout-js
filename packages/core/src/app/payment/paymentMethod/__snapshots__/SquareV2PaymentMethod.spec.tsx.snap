// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when using SquareV2PaymentMethod should render dummy form fields 1`] = `
<div
  data-test="squarev2_dummy_form"
  style={
    Object {
      "display": "none",
    }
  }
>
  <div
    className="form-field"
  >
    <div
      className="form-label optimizedCheckout-form-label"
      id="messageIsDefault"
    />
    <div
      className="form-input optimizedCheckout-form-input"
      id="inputIsDefault"
    />
  </div>
  <div
    className="form-field"
  >
    <div
      className="form-input optimizedCheckout-form-input form-input--focus optimizedCheckout-form-input--focus"
      id="inputIsFocus"
    />
  </div>
  <div
    className="form-field form-field--error"
  >
    <div
      className="form-inlineMessage"
      id="messageIsError"
    />
    <div
      className="form-input optimizedCheckout-form-input"
      id="inputIsError"
    />
  </div>
</div>
`;
